### PR TITLE
Training task # 5, Training task # 8 

### PR DESCRIPTION
**Pull request by hamza-56**
Training Task# 5: https://edlyio.atlassian.net/browse/EDE-1667
Training Task# 8: https://edlyio.atlassian.net/browse/EDE-1674

Deployed on: https://nameless-castle-31611.herokuapp.com/

- [x] Added api endpoint for user registration

- [x] Added api endpoints to get/ update user profiles

- [x] Added UserIsOwnerOrReadOnly permission to only allow user to update his profile

- [x] Custom Object Manager

- [x] Model Level Validations

- [x] Updated Postman collection

### CODE REVIEW

File: accounts/serializers.py [hunk 1]
```diff
@@ -1,16 +1,36 @@
...
         model = Log
         fields = '__all__'
+
+
+class UserSerializer(serializers.ModelSerializer):
+
+    password = serializers.CharField(write_only=True)
+
+    def create(self, validated_data):
+        return User.objects.create_user(**validated_data)
```
Comment (tasawernawaz) [inline accounts/serializers.py]: I believe you need to override the `create` method to save hashed password. Can you confirm this?

File: api/views.py [hunk 2]
```diff
@@ -1,35 +1,37 @@
...
+    permission_classes = (permissions.IsAdminUser, )
+    authentication_classes = (authentication.TokenAuthentication, )
 
     def get(self, request, format=None):
-        """
+        '''
         Return a list of all users.
-        """
+        '''
         usernames = [user.username for user in User.objects.all()]
```
Comment (tasawernawaz) [inline api/views.py]: can we use `<query>.values('username')`?

also, I would like to ask why we are only returning `usernames`?

File: api/views.py [hunk 3]
```diff
@@ -49,3 +51,25 @@ def get_queryset(self):
...
         user = self.request.user
         serializer.save(owner=user)
+
+
+class RegisterView(CreateAPIView):
+    model = User
+    serializer_class = UserSerializer
+
+
+class UserIsOwnerOrReadOnly(permissions.BasePermission):
```
Comment (tasawernawaz) [inline api/views.py]: move this to new file called `permissions.py`

File: accounts/serializers.py [hunk 4]
```diff
@@ -1,16 +1,36 @@
...
-            raise serializers.ValidationError(
-                f'Invalid timezone string {timezone_str}')
-        return timezone_str
-
     class Meta:
         model = Log
         fields = '__all__'
+
+
+class UserSerializer(serializers.ModelSerializer):
```
Comment (tasawernawaz) [inline accounts/serializers.py]: if this is only going to be user for registering users, rename it something like `UserCreateSerializer `

**Review by tasawernawaz — CHANGES_REQUESTED** [review]

File: accounts/serializers.py (hunk 1 above)
Comment (hamza-56) [reply accounts/serializers.py]: Yes, we override create to save hashed password.

Comment (tasawernawaz) [reply accounts/serializers.py]: kr dyn phir

**Review by tasawernawaz — APPROVED** [review]

**Pull request merged**